cmd/dra-example-kubeletplugin/driver.go

```diff
@@ -20,132 +20,111 @@ import (
+	client coreclientset.Interface
+	plugin kubeletplugin.DRAPlugin
 	state  *DeviceState
 }
 
 func NewDriver(ctx context.Context, config *Config) (*driver, error) {
+	driver := &driver{
+		client: config.coreclient,
+	}
+
 	state, err := NewDeviceState(config)
 	if err != nil {
 		return nil, err
 	}
+	driver.state = state
+
+	plugin, err := kubeletplugin.Start(
+		ctx,
+		driver,
+		kubeletplugin.KubeClient(config.coreclient),
+		kubeletplugin.NodeName(config.flags.nodeName),
+		kubeletplugin.DriverName(DriverName),
+		kubeletplugin.RegistrarSocketPath(PluginRegistrationPath),
+		kubeletplugin.PluginSocketPath(DriverPluginSocketPath),
+		kubeletplugin.KubeletPluginSocketPath(DriverPluginSocketPath))
+	if err != nil {
+		return nil, err
+	}
+	driver.plugin = plugin
 
-	d := &driver{
-		state: state,
+	var resources kubeletplugin.Resources
+	for _, device := range state.allocatable {
+		resources.Devices = append(resources.Devices, device)
 	}
+	plugin.PublishResources(ctx, resources)
 
-	return d, nil
+	return driver, nil
 }
 
 func (d *driver) Shutdown(ctx context.Context) error {
-	close(d.doneCh)
+	d.plugin.Stop()
 	return nil
 }
 
-func (d *driver) NodeListAndWatchResources(req *drapbv1.NodeListAndWatchResourcesRequest, stream drapbv1.Node_NodeListAndWatchResourcesServer) error {
-	model := d.state.getResourceModelFromAllocatableDevices()
-	resp := &drapbv1.NodeListAndWatchResourcesResponse{
-		Resources: []*resourceapi.ResourceModel{&model},
-	}
-
-	if err := stream.Send(resp); err != nil {
-		return err
-	}
-
-	//nolint:all,S1000: should use for range instead of for { select {} } (gosimple)
-	for {
-		select {
-		case <-d.doneCh:
-			return nil
-		}
-		// TODO: Update with case for when GPUs go unhealthy
-	}
-}
-
 func (d *driver) NodePrepareResources(ctx context.Context, req *drapbv1.NodePrepareResourcesRequest) (*drapbv1.NodePrepareResourcesResponse, error) {
 	klog.Infof("NodePrepareResource is called: number of claims: %d", len(req.Claims))
 	preparedResources := &drapbv1.NodePrepareResourcesResponse{Claims: map[string]*drapbv1.NodePrepareResourceResponse{}}
 
-	// In production version some common operations of d.nodeUnprepareResources
-	// should be done outside of the loop, for instance updating the CR could
-	// be done once after all HW was prepared.
 	for _, claim := range req.Claims {
-		preparedResources.Claims[claim.Uid] = d.nodePrepareResource(ctx, claim)
+		preparedResources.Claims[claim.UID] = d.nodePrepareResource(ctx, claim)
 	}
 
 	return preparedResources, nil
 }
 
 func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodePrepareResourceResponse {
-	if len(claim.StructuredResourceHandle) == 0 {
-		return &drapbv1.NodePrepareResourceResponse{
-			Error: "driver only supports structured parameters",
-		}
-	}
-
-	allocated, err := d.getAllocatedDevices(ctx, claim)
+	resourceClaim, err := d.client.ResourceV1alpha3().ResourceClaims(claim.Namespace).Get(
+		context.TODO(),
```

Comment:
If I change this I get the linting error:
```
[lint: cmd/dra-example-kubeletplugin/driver.go#L90](https://github.com/kubernetes-sigs/dra-example-driver/pull/50/files#annotation_24764568359)
Non-inherited new context, use function like `context.WithXXX` instead (contextcheck)
```

Reply:
I think there's been a mixup, perhaps you didn't save the file when you changed it to `ctx` locally? This linter warning appears exactly when a new `context.TODO()` context is created instead of reusing the `ctx` from arguments.

Here's my terminal log: I run linter - it's clean, and I show the file is changed like this thread's suggestion shows, then I checkout the file to original state with `context.TODO()` and voila, the linter complains about not inheriting argument's context, and creating a new one instead.
```bash
$ make lint                                                                                                                                                       
golangci-lint run ./...                                                                                                                                                                                            
byako@oleksiif-mobl:~/git/cdi/dra-example-driver $ git diff cmd/
diff --git a/cmd/dra-example-kubeletplugin/driver.go b/cmd/dra-example-kubeletplugin/driver.go
index ba795d8..9fde683 100644
--- a/cmd/dra-example-kubeletplugin/driver.go
+++ b/cmd/dra-example-kubeletplugin/driver.go
@@ -88,7 +88,7 @@ func (d *driver) NodePrepareResources(ctx context.Context, req *drapbv1.NodePrep
 
 func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodePrepareResourceResponse {
        resourceClaim, err := d.client.ResourceV1alpha3().ResourceClaims(claim.Namespace).Get(
-               context.TODO(),
+               ctx,
                claim.Name,
                metav1.GetOptions{})
        if err != nil {
byako@oleksiif-mobl:~/git/cdi/dra-example-driver $ git checkout HEAD cmd/dra-example-kubeletplugin/driver.go
Updated 1 path from ea67393
byako@oleksiif-mobl:~/git/cdi/dra-example-driver $ make lint
golangci-lint run ./...
cmd/dra-example-kubeletplugin/driver.go:90:87: Non-inherited new context, use function like `context.WithXXX` instead (contextcheck)
        resourceClaim, err := d.client.ResourceV1alpha3().ResourceClaims(claim.Namespace).Get(
                                                                                             ^
make: *** [Makefile:82: lint] Error 1
```